### ISSUE TITLE
When shall we update the pinned index state in gitlab release CI?

### ISSUE THREAD

**Issue by Mikolaj** [haskell/cabal#9819]
I'm opening a new separate ticket for this question, because this has cost us a serious problem in the 3.10.3.0 release process: an urgent serious upstream fix has not been included in a nearly released deemed-final binaries: https://github.com/haskell/tar/issues/89#issuecomment-2007386670

Older related discussion:

https://github.com/haskell/cabal/pull/9610#discussion_r1519344739

https://github.com/haskell/cabal/issues/9808

**Comment by Mikolaj** [comment]
@Kleidukos: in case of release 3.10.3.0 (and 3.12.0.0), due to https://github.com/haskell/tar/issues/89, I guess we have to update the index right now?

**Comment by Kleidukos** [comment]
Yes we have to restart the whole process. I'll submit a PR for 3.10

**Comment by hasufell** [comment]
Just fyi: the binaries that you test in the validate pipeline are potentially not the same as the ones you release (`cabal.project.validate` does not use index state pinning).

**Comment by Kleidukos** [comment]
@hasufell ok I'm addressing this in #9820

**Comment by Mikolaj** [comment]
@hasufell: yes, we are discussing this for the last week. For now, we `--dry-run` build the release project files with the pinned index state on GHA. There's a plan to run full tests on gitlab. There is a plan to use head.hackage to test against bleeding edge and then we could fix the index in the normal validate tests and not worry that our CI can't reproduce problems users (that compile from source) are having. I don't think we should make radical changes without a proper discussion, except for the disastrous too old index state on release pipeline.

**Comment by hasufell** [comment]
Well, I have no skin in this game. I have my own way of building cabal release binaries.

**Comment by chreekat** [comment]
Ouch. :(

What bit Cabal here is not the question of when to do semiautomatic freshness bumps, but the fact that the validation CI and the release CI don't share a pin. I think we can just call that a longstanding bug unrelated to the title of this issue.

Whether a pin is updated on a semiregular basis to keep deps fresh, or whether it's updated explicitly to pull in a specific required dep, all CI pipelines should use the same pin, and ideally it is only defined once!

**Comment by Kleidukos** [comment]
Isn't Phil trying to get a PR in that would substantially increase sharing between cabal project files?

**Comment by Kleidukos** [comment]
Anyhoo, I'll proceed with my PR for 3.10.

**Comment by Mikolaj** [comment]
> What bit Cabal here is not the question of when to do semiautomatic freshness bumps, but the fact that the validation CI and the release CI don't share a pin. I think we can just call that a longstanding bug unrelated to the title of this issue.

I don't follow. How sharing a pin would make us automatically use the fixed instead of the broken version of package `tar`?
 
> Whether a pin is updated on a semiregular basis to keep deps fresh, or whether it's updated explicitly to pull in a specific required dep, all CI pipelines should use the same pin, and ideally it is only defined once!

Thank you for your voice in the discussion. Actually it's already cited in https://github.com/haskell/cabal/issues/9808. How do you respond to the counter-arguments from the discussion, mostly the one that if we fix index state in all of our CI, we won't be seeing in our CI the errors users experience and report (some users compile cabal from source, others use Cabal as a library and so compile with new or even newest Hackage)? E.g., if newest `tar` was causing a visible error in our CI (it didn't) just as it did for the reporter, we wouldn't be seeing that error until after the release and so we would include a faulty `tar` (which had the deeper error for many releases, but not manifested and reported until the last one).

**Comment by philderbeast** [comment]
> Isn't Phil trying to get a PR in that would substantially increase sharing between cabal project files?

@Kleidukos #9565 is approved but I have some squashing to do before adding the merge label and one last look over my shoulder to see what has changed with projects on the master branch.

**Comment by hasufell** [comment]
> mostly the one that if we fix index state in all of our CI

Who said "all"?

**Comment by chreekat** [comment]
I confess I'm talking about strategy rather than tactics. From a strategic perspective, there is a tension between ensuring CI works reliably and ensuring any random use case of the package "just works". You can do both, to a degree, but it takes a lot of work. Which one do you prioritize?

> How do you respond to the counter-arguments from the discussion, mostly the one that if we fix index state in all of our CI, we won't be seeing in our CI the errors users experience and report (some users compile cabal from source, others use Cabal as a library and so compile with new or even newest Hackage)? 

How much work do you want to do, and when do you want to do it?

From my perspective, you should worry about having reliable CI first, so that new and existing contributors are able to contribute. Then you can start thinking about making more and more use cases "just work". For example, if people want to build cabal-install from source, you could try recommending a known-good dependency set. (Of course, this requires publishing the "recommended dependency set" somewhere.) If Cabal-the-library stops working with a newer index-state, then maybe Cabal's dependency bounds need revision.

Maybe this isn't useful advice, because I don't have any visibility into the kinds of issues users face when trying to build Cabal themselves. But again, which do you prioritize: reliable CI, or ensuring all possible use cases of the package "just work"? Right now Cabal seems to have neither, and focusing on one at a time might be wise.


> > mostly the one that if we fix index state in all of our CI
> 
> Who said "all"?

I did, for one.

**Comment by hasufell** [comment]
Well, GHCup doesn't need Cabal's release CI anymore. It would make no difference to me if you drop it altogether.

Of course, you might wish to release your own binaries regardless.

**Comment by Kleidukos** [comment]
@chreekat @Mikolaj starting with cabal.project.release and cabal.project.validate: what do you think?

**Comment by chreekat** [comment]
@Kleidukos seems fine, but I will leave it to the project maintainers' decision. :) I have other decisions I need to focus my limited energy on at this time.

**Comment by ffaf1** [comment]
I am still going through the discussion, but let me add this:

I am helping with releasing `Cabal 3.12`. During the release process, there is a step where you ask — after thorough testing — `hackage-security` maintainers to relax their bounds on `Cabal` and `Cabal-syntax`, and to release it on Hackage.

Of course after `hackage-security` was released on Hackage, I had to update `index-state:` in `cabal-project.release`.
